common/src/barretenberg_structures.rs

```diff
@@ -938,87 +968,81 @@ impl From<&Circuit> for ConstraintSystem {
-    let mut ql = Scalar::zero();
-    let mut qr = Scalar::zero();
-    let mut qo = Scalar::zero();
-
+    let mut cs = Constraint::default();
     // check mul gate
     if !gate.mul_terms.is_empty() {
         let mul_term = &gate.mul_terms[0];
-        qm = mul_term.0;
+        cs.qm = mul_term.0;
 
         // Get wL term
         let wL = &mul_term.1;
-        a = wL.witness_index() as i32;
+        cs.a = wL.witness_index() as i32;
 
         // Get wR term
         let wR = &mul_term.2;
-        b = wR.witness_index() as i32;
-    }
-
-    // If there is only one simplified fan term,
-    // then put it in qO * wO
-    // This is in case, the qM term is non-zero
-    if gate.linear_combinations.len() == 1 {
-        let qO_wO_term = &gate.linear_combinations[0];
-        qo = qO_wO_term.0;
-
-        let wO = &qO_wO_term.1;
-        c = wO.witness_index() as i32;
+        cs.b = wR.witness_index() as i32;
     }
 
-    // XXX: This is a code smell. Refactor to be better. Maybe change Barretenberg to take vectors
-    // If there is more than one term,
-    // Then add normally
-    if gate.linear_combinations.len() == 2 {
-        let qL_wL_term = &gate.linear_combinations[0];
-        ql = qL_wL_term.0;
-
-        let wL = &qL_wL_term.1;
-        a = wL.witness_index() as i32;
-
-        let qR_wR_term = &gate.linear_combinations[1];
-        qr = qR_wR_term.0;
-
-        let wR = &qR_wR_term.1;
-        b = wR.witness_index() as i32;
-    }
-
-    if gate.linear_combinations.len() == 3 {
-        let qL_wL_term = &gate.linear_combinations[0];
-        ql = qL_wL_term.0;
-
-        let wL = &qL_wL_term.1;
-        a = wL.witness_index() as i32;
-
-        let qR_wR_term = &gate.linear_combinations[1];
-        qr = qR_wR_term.0;
-
-        let wR = &qR_wR_term.1;
-        b = wR.witness_index() as i32;
-
-        let qO_wO_term = &gate.linear_combinations[2];
-        qo = qO_wO_term.0;
-
-        let wO = &qO_wO_term.1;
-        c = wO.witness_index() as i32;
+    for term in &gate.linear_combinations {
+        cs.set_linear_term(term.0, term.1.witness_index() as i32);
     }
 
     // Add the qc term
-    let qc = gate.q_c;
-
-    Constraint {
-        a,
-        b,
-        c,
-        qm,
-        ql,
-        qr,
-        qo,
-        qc,
+    cs.qc = gate.q_c;
+    cs
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::barretenberg_structures::serialize_arithmetic_gates;
+    use acvm::acir::native_types::{Expression, Witness};
+    use acvm::FieldElement as Scalar;
+
+    #[test]
+    fn serialize_expression() {
+        let x1 = Witness::new(1);
+        let x3 = Witness::new(3);
+        let two = Scalar::one() + Scalar::one();
+        let e = Expression {
+            mul_terms: vec![(Scalar::one(), x1, x1)],
+            linear_combinations: vec![(two, x1), (-Scalar::one(), x3)],
+            q_c: Scalar::one(),
+        };
+        let constrain = serialize_arithmetic_gates(&e);
+        assert!(constrain.a == 1);
```

Comment:
```suggestion
        assert_eq!(constrain.a, 1);
```

and all other instances.

Reply:
done